smartsim/settings/dispatch.py

```diff
@@ -0,0 +1,169 @@
+        to_launcher: type[LauncherLike[_T]],
+        allow_overwrite: bool = False,
+    ) -> t.Callable[[type[LaunchArgBuilder[_T]]], type[LaunchArgBuilder[_T]]] | None:
+        """A type safe way to add a mapping of settings builder to launcher to
+        handle the settings at launch time.
+        """
+
+        def register(
+            args_: type[LaunchArgBuilder[_T]], /
+        ) -> type[LaunchArgBuilder[_T]]:
+            if args_ in self._dispatch_registry and not allow_overwrite:
+                launcher_type = self._dispatch_registry[args_]
+                raise TypeError(
+                    f"{args_.__name__} has already been registered to be "
+                    f"launched with {launcher_type}"
+                )
+            self._dispatch_registry[args_] = to_launcher
+            return args_
+
+        if args is not None:
+            register(args)
+            return None
+        return register
+
+    def get_launcher_for(
+        self, args: LaunchArgBuilder[_T] | type[LaunchArgBuilder[_T]], /
+    ) -> type[LauncherLike[_T]]:
+        """Find a type of launcher that is registered as being able to launch
+        the output of the provided builder
+        """
+        if not isinstance(args, type):
+            args = type(args)
+        launcher_type = self._dispatch_registry.get(args, None)
+        if launcher_type is None:
+            raise TypeError(
+                f"{type(self).__name__} {self} has no launcher type to "
+                f"dispatch to for argument builder of type {args}"
+            )
+        # Note the sleight-of-hand here: we are secretly casting a type of
+        # `LauncherLike[Any]` to `LauncherLike[_T]`. This is safe to do if all
+        # entries in the mapping were added using a type safe method (e.g.
+        # `Dispatcher.dispatch`), but if a user were to supply a custom
+        # dispatch registry or otherwise modify the registry THIS IS NOT
+        # NECESSARILY 100% TYPE SAFE!!
+        return launcher_type
+
+
+default_dispatcher: t.Final = Dispatcher()
+
+
+# >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
+# TODO: move these to a common module under `smartsim._core.launcher`
+# -----------------------------------------------------------------------------
+
+
+def create_job_id() -> LaunchedJobID:
+    return LaunchedJobID(str(uuid.uuid4()))
+
+
+class LauncherLike(t.Protocol[_T_contra]):
+    def start(self, launchable: _T_contra) -> LaunchedJobID: ...
+    @classmethod
+    def create(cls, exp: Experiment) -> Self: ...
+
+
+class ShellLauncher:
+    """Mock launcher for launching/tracking simple shell commands"""
+
+    def __init__(self) -> None:
+        self._launched: dict[LaunchedJobID, sp.Popen[bytes]] = {}
+
+    def start(self, launchable: t.Sequence[str]) -> LaunchedJobID:
+        id_ = create_job_id()
+        exe, *rest = launchable
+        self._launched[id_] = sp.Popen((helpers.expand_exe_path(exe), *rest))
```

Comment:
the variable self._launched[id_] : what is this?

Reply:
Right now, this is just creating a unique id for the job, opening a suprocess to start the job, and then adding the id and the job that it corresponds to to some mapping maintained by the launcher.

We've not fully thought through  the API, but in theory this will allow launchers to keep track of which jobs they have started, and thus which jobs they are capable of checking the statuses for, which might lead to patterns like this:
```py
>>> shell_launcher = ShellLauncher()
>>> job_id = shell_launcher.start(["echo", "hello", "world"])
>>> print(shell_launcher.status(job_id))
RUNNING
```
where `ShellLauncher.status` uses that underlying mapping

Reply:
But for right now, it is just a forever growing collection that we do not read from (aka, just a fancy memory leak, hahaha)